tests/Cms/Users/UserTest.php

```diff
@@ -239,4 +239,87 @@ public function testUserModel()
 
         User::$models = [];
     }
+    
+    public function testLoginBeforeHook()
+    {
+        $phpUnit = $this;
+        
+        $app = new App([
+            'users' => [
+                ['email' => 'test@getkirby.com']
+            ],
+            'hooks' => [
+                'user.login:before' => function ($user, $session) use ($phpUnit) {
+                    $phpUnit->assertEquals('test@getkirby.com', $user->email());
+                    $phpUnit->assertEquals($session, S::instance());
+                }
+            ]
+        ]);
+
+        $user = $app->user('test@getkirby.com');
+
+        $app->trigger('user.login:before', $user, $app->session());
```

Comment:
The tests currently don't really test the new functionality of the login and logout methods, but just the hook itself – which is more of a test of the `$app->trigger()` method. Maybe you can find a way to trigger and test the hooks when calling `$user->loginPasswordless()` and `$user->logout()` directly – you could even test the before and after hooks at the same time that way.

Reply:
Yes, it was a self-test method 😄 Updated with your said as called directly the methods.

Reply:
I'd combine the test into two test methods – one for login and one for logout. Each one would test both hooks of the respective method.